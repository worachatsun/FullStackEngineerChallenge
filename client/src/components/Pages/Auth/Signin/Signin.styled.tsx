import styled from "styled-components";

export const Container = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
`;

export const ErrorMessage = styled.div`
    display: flex;
    justify-content: center;
    background: #ff0033;
    opacity: 0.9;
    color: white;
    margin-top: 20px;
`;

export const ButtonContainer = styled.div`
    display: flex;
    justify-content: flex-end;
    margin-top: 20px;
`;
