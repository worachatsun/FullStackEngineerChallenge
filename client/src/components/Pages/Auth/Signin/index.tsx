import React, { FunctionComponent, useContext } from "react";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { SIGNIN_API } from "../../../../constants/routes";
import { UserActionType, UserContext } from "../../../../context/UserContext";
import Button from "../../../commons/Button";
import Input from "../../../commons/Input";
import { HttpMethod, mutator } from "../../../commons/utils/client";
import { Container, Label } from "./Signin.styled";
import { yupResolver } from "@hookform/resolvers";
import { SigninSchema } from "../../../../schemas/signin";

interface ISignin {
    username: string;
    password: string;
}

const Signin: FunctionComponent = () => {
    const { register, handleSubmit, errors } = useForm<ISignin>({ resolver: yupResolver(SigninSchema) });
    const { dispatch } = useContext(UserContext);
    const history = useHistory();

    const onSubmit = async (user: ISignin) => {
        try {
            const { username, password } = user;
            const { response } = await mutator(SIGNIN_API, HttpMethod.POST, undefined, { username, password });
            if (!response?.ok) throw new Error(`${response?.status}: ${response?.statusText}`);
            const data = await response?.json();
            localStorage.setItem("token", data.user.token);
            dispatch({ type: UserActionType.ADD_USER, payload: data.user.user });
            history.push("/list");
        } catch (error) {
            console.error(error);
        }
    };

    return (
        <Container>
            <form onSubmit={handleSubmit(onSubmit)}>
                <Label>Signin</Label>
                <Input title={"Username"} name="username" errors={errors.username} register={register} />
                <Input
                    title={"Password"}
                    name="password"
                    errors={errors.password}
                    register={register}
                    type={"password"}
                />
                <Button>Signin</Button>
            </form>
        </Container>
    );
};

export default Signin;
