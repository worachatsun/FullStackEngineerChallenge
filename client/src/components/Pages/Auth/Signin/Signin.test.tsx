import Signin from "./index";
import React from "react";
import renderer from "react-test-renderer";

describe("<Signin />", () => {
    test("should render correctly", () => {
        const render = renderer.create(<Signin />).toJSON();
        expect(render).toMatchSnapshot();
    });
});
