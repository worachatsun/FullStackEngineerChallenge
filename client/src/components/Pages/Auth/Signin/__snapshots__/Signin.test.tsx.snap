// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signin /> should render correctly 1`] = `
<div
  className="sc-AxiKw jAHgLp"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="Username"
      >
        Username
        :
        <input
          className="sc-AxirZ lmasuR"
          name="username"
          type="text"
        />
      </label>
    </div>
    <div>
      <label
        htmlFor="Password"
      >
        Password
        :
        <input
          className="sc-AxirZ lmasuR"
          name="password"
          type="password"
        />
      </label>
    </div>
    <button
      className="sc-AxjAm fbqdBC"
    >
      Signin
    </button>
  </form>
</div>
`;
